File: Problems/No.101_Symmetric_Tree/symmetric_tree_iterative.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        
        # parameters
        check = [[root.left, root.right]]
        
        # start compute
        while (check!=[]):
            
            left, right = check[0]
            del check[0]
            
            if (left==None and right==None):
                continue
            elif ((not left and right) or (left and not right)):
                return False
            elif (left.val != right.val):
                return False
            else:
                
                check.append([left.right, right.left])
                check.append([left.left, right.right])
                
        return True
            
        

File: Problems/No.101_Symmetric_Tree/symmetric_tree_iterative.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        
        // parameters
        TreeNode *left, *right;
        queue<TreeNode*> check_left, check_right;
        
        // start compute, using queues
        check_left.push(root->left);
        check_right.push(root->right);
        while (!check_left.empty() and !check_right.empty()) {
            
            left = check_left.front();
            check_left.pop();
            right = check_right.front();
            check_right.pop();
            
            if (left==NULL and right==NULL) continue;
            else if ((!left and right) or (left and !right)) return false;
            else if (left->val != right->val) return false;
            else {
                
                check_left.push(left->left);
                check_right.push(right->right);
                check_left.push(left->right);
                check_right.push(right->left);
                
            }
            
        }
        
        return true;
        
    }
};
